__tests__/api/BundleContent/Bundle.test.ts

```diff
@@ -152,8 +171,186 @@ describe("Bundle01", () => {
+        const spy2 = jest.spyOn(fs, "accessSync").mockImplementationOnce(() => { throw new Error("Wibble"); });
+
+        let err: Error;
+        try {
+          bund.prepareForSave();
+        }
+        catch (error) {
+          err = error;
+        }
+
+        expect(err).toBeDefined();
+        expect(err.message).toMatchSnapshot();
+    });
+    it("should complain if no write permission to META-INF directory", () => {
+
+        // Create a Bundle
+        const bund = new Bundle("__tests__/__resources__/ExampleBundle01", false, false);
+
+        const spy1 = jest.spyOn(fs, "existsSync").mockImplementationOnce(() => ( true ));
+        const spy2 = jest.spyOn(fs, "accessSync").mockImplementationOnce(() => { throw new Error("Wibble"); });
+
+        let err: Error;
+        try {
+          bund.prepareForSave();
+        }
+        catch (error) {
+          err = error;
+        }
+
+        expect(err).toBeDefined();
+        expect(err.message).toMatchSnapshot();
+    });
+    it("should complain if no overwrite permission to manifest", () => {
+
+        // Create a Bundle
+        const bund = new Bundle("__tests__/__resources__/ExampleBundle01", false, false);
+
+        const spy1 = jest.spyOn(fs, "existsSync").mockImplementationOnce(() => ( true ));
+        const spy2 = jest.spyOn(fs, "accessSync").mockImplementationOnce(() => ( true ));
+        const spy3 = jest.spyOn(fs, "existsSync").mockImplementationOnce(() => ( true ));
+
+        let err: Error;
+        try {
+          bund.prepareForSave();
+        }
+        catch (error) {
+          err = error;
+        }
+
+        expect(err).toBeDefined();
+        expect(err.message).toMatchSnapshot();
+    });
+    it("should complain if no write permission to manifest", () => {
+
+        // Create a Bundle
+        const bund = new Bundle("__tests__/__resources__/ExampleBundle01", false, true);
+
+        const spy1 = jest.spyOn(fs, "existsSync").mockImplementationOnce(() => ( true ));
+        const spy2 = jest.spyOn(fs, "accessSync").mockImplementationOnce(() => ( true ));
+        const spy3 = jest.spyOn(fs, "existsSync").mockImplementationOnce(() => ( true ));
+        const spy4 = jest.spyOn(fs, "accessSync").mockImplementationOnce(() => { throw new Error("Wibble"); });
+
+        let err: Error;
+        try {
+          bund.prepareForSave();
+        }
+        catch (error) {
+          err = error;
+        }
+
+        expect(err).toBeDefined();
+        expect(err.message).toMatchSnapshot();
+    });
+    it("should complain if can't make a new META-INF directory", () => {
+
+        // Create a Bundle
+        const bund = new Bundle("__tests__/__resources__/ExampleBundle01", false, true);
+
+        const spy1 = jest.spyOn(fs, "existsSync").mockImplementationOnce(() => ( true ));
+        const spy2 = jest.spyOn(fs, "accessSync").mockImplementationOnce(() => ( true ));
```

Comment:
I'm concerned using this approach will lead to brittle tests that are tightly coupled to the implementation. E.g. reordering of the accessSync() calls in the implementation would break this test. Suggest rather than repeated calls to mockImplementationOnce() for the same function you could do something like:
```()
jest.spyOn(fs, "accessSync").mockImplementationOnce((path: string) => {
  if (path === "__tests__/__resources__/ExampleBundle01") {
    return true;
  if (path === "__tests__/__resources__/ExampleBundle01/META-INF") {
    return false;
  }
  return true;
});
```
I think this expresses more clearly what you're trying to test as well. It should also help avoid unexpected interactions with other bits of code that use the filesystem.

Reply:
The big issue I had whilst writing this involved cascading errors (a mock for an earlier test interfered with the processing of a later test). The fix to that was to use mockImplementationOnce() and to explicitly check the count for each spy... it was only then that the tests became sufficiently isolated from each other that a failure in one didn't cause a cascade of failures in the others. So far I've found the combination to work quite well, if the implementation changes then the only tests to fail are those that will need updating.

Unfortunately the suggested alternative wont work, at least not in all cases; there are some paths through the code where we need different interactions with the same resource to return different answers; for example, the first interaction with the manifest claims to work, but the second interaction throws an exception.

Reply:
I think this warrants further investigation, I'll raise an issue to follow this up.